src/foraging_gui/Foraging.py

```diff
@@ -857,72 +921,81 @@ def _TrainingStage(self):
-                    try: # load the paramter used by last trial
-                        value=np.array([self.TrainingStagePar[Task][CurrentTrainingStage][key]])
-                        Tag=0
-                    # sometimes we only have training parameters, no behavior parameters
-                    except Exception as e:
-                        logging.error(str(e))
-                        value=self.TrainingStagePar[Task][CurrentTrainingStage][key]
-                        Tag=1
-                    if isinstance(widget, QtWidgets.QPushButton):
-                        pass
-                    if type(value)==bool:
-                        Tag=1
-                    else:
-                        if len(value)==0:
-                            value=np.array([''], dtype='<U1')
-                            Tag=0
-                    if type(value)==np.ndarray:
-                        Tag=0
-                    if isinstance(widget, QtWidgets.QLineEdit):
-                        if Tag==0:
-                            widget.setText(value[-1])
-                        elif Tag==1:
-                            widget.setText(value)
-                    elif isinstance(widget, QtWidgets.QComboBox):
-                        if Tag==0:
-                            index = widget.findText(value[-1])
-                        elif Tag==1:
-                            index = widget.findText(value)
-                        if index != -1:
-                            widget.setCurrentIndex(index)
-                    elif isinstance(widget, QtWidgets.QDoubleSpinBox):
-                        if Tag==0:
-                            widget.setValue(float(value[-1]))
-                        elif Tag==1:
-                            widget.setValue(float(value))
-                    elif isinstance(widget, QtWidgets.QSpinBox):
-                        if Tag==0:
-                            widget.setValue(int(value[-1]))
-                        elif Tag==1:
-                            widget.setValue(int(value))
-                    elif isinstance(widget, QtWidgets.QTextEdit):
-                        if Tag==0:
-                            widget.setText(value[-1])
-                        elif Tag==1:
-                            widget.setText(value)
-                    elif isinstance(widget, QtWidgets.QPushButton):
-                        if key=='AutoReward':
-                            if Tag==0:
-                                widget.setChecked(bool(value[-1]))
-                            elif Tag==1:
-                                widget.setChecked(value)
-                            self._AutoReward()
-                else:
-                    widget = widget_dict[key]
-                    if not (isinstance(widget, QtWidgets.QComboBox) or isinstance(widget, QtWidgets.QPushButton)):
-                        widget.clear()
+                self._set_parameters(key,widget_dict,self.TrainingStagePar[Task][CurrentTrainingStage])
```

Comment:
I use three parameters `key`, `widget_dict` and `parameters` here, and use a for loop to iterate through all the keys on purpose. In this way, I can exclude some keys from the widget_dict. There are multiple ways to do it. It's more relevant to encapsulation of for loops before or after.

The training stage is still useful even the training parameters are determined by autotrain. In some cases, we still need to explore some parameters in experiments, and we can use the training stage parameters. 

In autotrain, you have frozen the training phase parameters, and it is safe.

Reply:
Regarding point 4, I agree with you. I probably won't change it here, but will in a future update.

Reply:
> I use three parameters `key`, `widget_dict` and `parameters` here, and use a for loop to iterate through all the keys on purpose. In this way, I can exclude some keys from the widget_dict. There are multiple ways to do it. It's more relevant to encapsulation of for loops before or after.

I know your code works. I'm just proposing maybe a more intuitive and clear way of encapsulating the `set_parameter` function. It's not about whether you use for loops before or after, but about which list you apply the for loops on. I suggest you run the for loop on `para_dict`, where you explicitly list what parameters you want to change, instead of looping over all widgets in the `TrainingParameters` Qt Group and exclude those parameters you don't want. I know both ways do the same job, but the former looks more reusable, concise, and robust. For example, as I mentioned, the way you do a for loop over the children of `TrainingParameters` makes your code depend on the GUI configuration, which sounds less robust to me.

> The training stage is still useful even the training parameters are determined by autotrain. In some cases, we still need to explore some parameters in experiments, and we can use the training stage parameters.
> In autotrain, you have frozen the training phase parameters, and it is safe.

In my suggestion, I did not exclude the use of the training stage:
>This dictionary could come from a full task schema (like your _**self.TrainingStagePar[Task][CurrentTrainingStage]**_ and my automatic training stage)

Reply:
I think that you are not satisfied with the internal organization of `_set_parameters` instead of encapsulation of for loops outside or in the `_set_parameters`. Let's optimize it in a separate PR. There are a lot of stuff going on in this branch.